mitodl/ocw-course-hugo-theme
# docker configuration

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-course-hugo-theme/issues/29

#### What's this PR do?
This PR adds a `docker` folder with the relevant configuration files to build a docker image of this application and run it for local development.

#### How should this be manually tested?
You should have a local copy of `ocw-to-hugo` to generate content for various steps to test this PR.  Read the README and the example configuration file.  Start with an empty `.env`, placed inside the `docker` folder and then do the following:

 - Use `ocw-to-hugo` in download mode to download and generate content for a course, using the `private` folder as a staging area.  For example: `node . -i private/input -o private/output -c private/courses.json --download`.  Make sure you have AWS credentials configured for access to an `open-learning-course-data` bucket and you have a single course listed in your `private/courses.json` file like: `{"courses":["18-06-linear-algebra-spring-2010"]}`.
 - Back in `ocw-course-hugo-theme`, `cd docker`
 - `docker-compose build`
 - `docker-compose up` - This should give you an empty webpage at http://localhost:1313
 - Using the course we downloaded earlier, copy the contents of the course's content folder (i.e. `ocw-to-hugo/private/output/content/courses/18-06-linear-algebra-spring-2010/*`) to `docker/hugo/content` and copy the course's data template (i.e. `ocw-to-hugo/private/output/data/courses/18-06-linear-algebra-spring-2010.json`) to `docker/hugo/data` and rename it to `course.json`.  Reload the page at http://localhost:1313 and you should be able to browse the course pages.  Afterwards, stop the dev server in your terminal with ctrl + c and remove the data we copied into `docker/hugo/content` and `docker/hugo/data`.
 - Copy the contents of `docker/example.env` to `docker/.env` and alter `OCW_TO_HUGO_PATH` to match your path to `ocw-to-hugo`.  Set `OCW_TO_HUGO_DOWNLOAD` to `false` and set `OCW_TO_HUGO_INPUT` to `/ocw-to-hugo/private/input`.  Set `OCW_COURSE_ID` to the `course_id` of the course you downloaded in step 1 (i.e. `18-06-linear-algebra-spring-2010`).  Run `docker-compose up` and ensure that you can browse the same course site you did earlier at http://localhost:1313.
 - Make sure you have AWS credentials configured in your shell in the same way you would for `ocw-to-hugo`.  Then, set `OCW_TO_HUGO_DOWNLOAD` to `true` and remove the `OCW_TO_HUGO_INPUT` variable or set it to `/ocw-data/input`.  Set `OCW_TEST_COURSE` to a course we haven't used yet, for example, `8-04-quantum-physics-i-spring-2016`.  Run `docker-compose up` and visit http://localhost:1313.  You should see the new course there.
 - Test hot reloading of the node dev server by making a change to any of the CSS and making sure it properly reloads with your changes.  For example, in `header.scss` if you change the `background-color` of `#course-banner` it should be immediately apparent: 
```
#course-banner {
  color: $white;
  background-color: $course-blue;

  a {
    text-decoration: none;
  }
}
```
 - Test hot reloading of the Hugo server by making a change to any templates, for example,  `course_home.html` if you're looking at the course home page.  You should see the page refresh and your change rendered.
